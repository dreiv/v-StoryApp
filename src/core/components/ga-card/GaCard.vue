<script lang="ts" setup>
export interface GaCardProps {
  title?: string
}

defineProps<GaCardProps>()
</script>

<template>
  <div :class="$style.card">
    <header v-if="$slots.header || title" :class="$style.header">
      <slot name="header">
        <slot name="title">
          <span :class="$style.title">{{ title }}</span>
        </slot>

        <div v-if="$slots.actions" :class="$style.actions">
          <slot name="actions" />
        </div>
      </slot>
    </header>

    <div v-if="$slots.default"><slot /></div>

    <div v-if="$slots.interactive" :class="$style.interactive">
      <slot name="interactive" />
    </div>
  </div>
</template>

<style module>
.card {
  display: flex;
  flex-direction: column;
  gap: var(--ga-size-spacing-03);

  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-primary);
  border-radius: var(--ga-radius);
  padding: var(--ga-size-spacing-05);
}

.header {
  display: flex;
  justify-content: space-between;
  gap: var(--ga-size-spacing-05);
}

.title {
  font-weight: var(--ga-font-weight-bold);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  letter-spacing: var(--ga-text-md-tracking);
}

.actions {
  display: flex;
  align-items: center;
  gap: var(--ga-size-spacing-05);
}

.interactive {
  display: flex;
  justify-content: end;
}
</style>
