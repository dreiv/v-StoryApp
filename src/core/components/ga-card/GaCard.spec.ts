import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaCard from './GaCard.vue'

describe('GaCard', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = shallowMount(GaCard)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('should render correctly with default props', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render title when provided', () => {
    wrapper = shallowMount(GaCard, { props: { title: 'Card Title' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.title').text()).toBe('Card Title')
  })

  it('should render description when provided', () => {
    wrapper = shallowMount(GaCard, { props: { description: 'Card description' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.description').text()).toBe('Card description')
  })

  it('should render checkbox when selectable is true', () => {
    wrapper = shallowMount(GaCard, { props: { selectable: true } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('input[type="checkbox"]').exists()).toBe(true)
  })

  it('should render chevron when expandable is true', () => {
    wrapper = shallowMount(GaCard, { props: { expandable: true } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.chevron').exists()).toBe(true)
  })

  it('should render top slot when provided', () => {
    wrapper = shallowMount(GaCard, {
      slots: {
        topSlot: '<div class="test-top-slot">Top slot</div>',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.test-top-slot').exists()).toBe(true)
  })

  it('should render bottom slot when provided', () => {
    wrapper = shallowMount(GaCard, {
      slots: {
        bottomSlot: '<div class="test-bottom-slot">Bottom slot</div>',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.test-bottom-slot').exists()).toBe(true)
  })

  it('should render bottom slot when expanded is true', () => {
    wrapper = shallowMount(GaCard, {
      props: { expandable: true, expanded: true },
      slots: {
        bottomSlot: '<div class="test-bottom-slot">Bottom slot</div>',
      },
    })
    expect(wrapper.find('.bottomSlot').exists()).toBe(true)
    expect(wrapper.find('.test-bottom-slot').exists()).toBe(true)
  })

  it('should render interactive icon slot when provided', () => {
    wrapper = shallowMount(GaCard, {
      slots: {
        interactiveIcon: '<div class="test-icon">Icon</div>',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.test-icon').exists()).toBe(true)
  })

  it('should not emit click event when disabled', async () => {
    wrapper = shallowMount(GaCard, { props: { disabled: true } })
    await wrapper.trigger('click')
    expect(wrapper.emitted('click')).toBeFalsy()
  })

  it('should prevent event propagation when chevron is clicked', async () => {
    wrapper = shallowMount(GaCard, { props: { expandable: true } })
    const stopPropagationSpy = vi.fn()
    await wrapper.find('.chevron').trigger('click', { stopPropagation: stopPropagationSpy })
    expect(stopPropagationSpy).toHaveBeenCalled()
  })
})
