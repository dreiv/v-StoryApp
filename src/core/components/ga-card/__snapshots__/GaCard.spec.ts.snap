// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaCard > should render bottom slot when provided 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <div class="bottomSlot">
    <div class="test-bottom-slot">Bottom slot</div>
  </div>
</div>"
`;

exports[`GaCard > should render checkbox when selectable is true 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <div class="selector"><input type="checkbox" class="checkbox"></div>
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;

exports[`GaCard > should render chevron when expandable is true 1`] = `
"<details class="card">
  <summary class="summary">
    <!-- Top slot (optional) - #8 -->
    <!--v-if-->
    <div class="cardContent">
      <!-- Selector (optional) - #6 -->
      <!--v-if-->
      <div class="mainContent">
        <!-- Title - #2 -->
        <!--v-if-->
        <!-- Description - #3 -->
        <!--v-if-->
        <!-- Default slot for custom content -->
        <!--v-if-->
      </div>
      <div class="rightSection">
        <!-- Interactive icon (optional) - #5 -->
        <!--v-if-->
        <!-- Expandable chevron icon is automatically handled by the details element -->
        <div class="chevron"><svg class="chevronIcon" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" width="16" height="16">
            <path d="M6 9L12 15L18 9" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
          </svg></div>
      </div>
    </div>
  </summary><!-- Bottom slot (optional) - #7 -->
  <div class="bottomSlot"></div>
</details>"
`;

exports[`GaCard > should render correctly with default props 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;

exports[`GaCard > should render description when provided 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <div class="description">Card description</div><!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;

exports[`GaCard > should render interactive icon slot when provided 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <div class="interactiveIcon">
        <div class="test-icon">Icon</div>
      </div>
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;

exports[`GaCard > should render title when provided 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <!--v-if-->
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <div class="title">Card Title</div><!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;

exports[`GaCard > should render top slot when provided 1`] = `
"<!-- Non-expandable version -->
<div class="card">
  <!-- Top slot (optional) - #8 -->
  <div class="topSlot">
    <div class="test-top-slot">Top slot</div>
  </div>
  <div class="cardContent">
    <!-- Selector (optional) - #6 -->
    <!--v-if-->
    <div class="mainContent">
      <!-- Title - #2 -->
      <!--v-if-->
      <!-- Description - #3 -->
      <!--v-if-->
      <!-- Default slot for custom content -->
      <!--v-if-->
    </div>
    <div class="rightSection">
      <!-- Interactive icon (optional) - #5 -->
      <!--v-if-->
    </div>
  </div><!-- Bottom slot for non-expandable cards - #7 -->
  <!--v-if-->
</div>"
`;
